ndk/src/midi.rs

```diff
@@ -0,0 +1,271 @@
+    fn from_ptr(inner: NonNull<ffi::AMidiInputPort>) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiInputPort {
+        self.inner.as_ptr()
+    }
+
+    /// Sends data to the specified input port.
+    pub fn send(&self, buffer: &[u8]) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_send(self.as_ptr(), buffer.as_ptr(), buffer.len() as ffi::size_t)
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+
+    /// Sends a message with a 'MIDI flush command code' to the specified port.
+    ///
+    /// This should cause a receiver to discard any pending MIDI data it may have accumulated and
+    /// not processed.
+    pub fn send_flush(&self) -> Result<()> {
+        let result = unsafe { ffi::AMidiInputPort_sendFlush(self.as_ptr()) };
+        NdkMediaError::from_status(result)
+    }
+
+    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_sendWithTimestamp(
+                self.as_ptr(),
+                buffer.as_ptr(),
+                buffer.len() as ffi::size_t,
+                timestamp,
+            )
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+}
+
+impl<'a> Drop for MidiInputPort<'a> {
+    fn drop(&mut self) {
+        unsafe { ffi::AMidiInputPort_close(self.as_ptr()) };
+    }
+}
+
+pub struct MidiOutputPort<'a> {
+    inner: NonNull<ffi::AMidiOutputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiOutputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiOutputPort")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<'a> MidiOutputPort<'a> {
+    fn from_ptr(inner: NonNull<ffi::AMidiOutputPort>) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiOutputPort {
+        self.inner.as_ptr()
+    }
+
+    /// Receives the next pending MIDI message.
+    ///
+    /// To retrieve all pending messages, the client should repeatedly call this method until it
+    /// returns `Some(_)`.
+    ///
+    /// Note that this is a non-blocking call. If there are no Midi messages are available, the
+    /// function returns `None` immediately (for 0 messages received).
+    pub fn receive(&self, buffer: &mut [u8]) -> Result<(MidiOpcode, i64)> {
```

Comment:
I'd ask to turn these monospace doc-comments into intradoc links, but also change them to `Ok`/`Err` to match the `Result` here.

Reply:
Fixed!